src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs

```diff
@@ -282,5 +287,43 @@ internal static void VerifyOverridesPrintMembersFromBase(MethodSymbol overriding
                 diagnostics.Add(ErrorCode.ERR_DoesNotOverrideBaseMethod, overriding.Locations[0], overriding, baseType);
             }
         }
+
+        internal static bool AreAllPrintablePropertyGettersReadOnly(NamedTypeSymbol containingType)
+        {
+            return !containingType.GetMembers().Any(m => hasNonReadOnlyGetter(m));
```

Comment:
Calling `containingType.GetMembers()` here (executed in base constructor) causes `StackOverflowException` :(

Reply:
The `SynthesizedRecordPrintMembers` is constructed when the members of the record are created (in `SourceMemberContainerTypeSymbol`). So we cannot ask about the members of the record in that process...

The way we've sorted this out in the past is that the logic in `SourceMemberContainerTypeSymbol` is phased. We create all the declared members first (those in source) then we create the synthesized ones. The synthesized ones depend on the declared ones, but not vice-versa. This avoids circularity.

I think something like that would work for your purpose: looks at all the declared properties without pulling on all members. That probably needs to be done in `AddSynthesizedRecordMembersIfNecessary` which has access to all declared members and instantiates the `SynthesizedRecordPrintMembers`.

Reply:
Thank you, passing members into constructor from  AddSynthesizedRecordMembersIfNecessary worked. With current implementation we are interested only in user defined properties(since they could have non readonly getters), but it would, probably, be more reliable to check all of them.
I've updated code. I'll check other tests and publish PR if everything is good.

Reply:
It looks like things are going in the right direction. 

`TopLevelEditingTests.RecordStruct_ImplementSynthesized_PrintMembers` should probably make the explicit implementation use the 'readonly' modifier so that EnC doesn't see it as a rude edit.

`RecordStructTests.Deconstruct_Empty_WithParameterList` should probably update baselines.

`RecordStructTests.ToString_TopLevelRecord_OneField_ConstrainedValueType` shows a bug though.
```

Reply:
I took another pass over the spec with the complexity of possibly having to call non-readonly methods on fields in mind. A few observations:

1. for Equals and GetHashCode the spec change is correct as written. We always make a value copy of each field and call into `EqualityComparer<TField>.Default.Equals()/GetHashCode` (which simply calls back into the `TField`'s Equals/GetHashCode if the field's value is not null.). Making the method `readonly` offers a benefit because if the user ever has a readonly ref to the record struct, we won't make a copy of the struct before the call. Therefore fewer copies are made overall.

2. for `PrintMembers()`/`ToString()` it's a little harder to feel out whether the spec change is correct, especially in light of tests like `ToString_TopLevelRecord_OneField_ConstrainedValueType`. However, after some thought I've decided that the spec as written is still the right thing, because it's *already* the behavior for a PrintMembers contained in a readonly record struct. If you inspect the PrintMembers method here, you'll see that the field gets copied the same way: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4FgBQABAzAAgCcBTCBAexgBsBPI4gY3MIXwGcAXQgVwY/wBKjADwAVAHz4A7gAtiJfKPwh2XXhywBvLPl3489MpVqL8AMyjEqCANxYAvliA=). This is only an issue for the very particular case of fields whose type is a type parameter constrained to struct. For unconstrained type parameters, we simply box and pass to the StringBuilder as an `object`.

To summarize: I think the spec change proposed in dotnet/csharplang#4890 is correct, and the baseline change in `RecordStructTests.ToString_TopLevelRecord_OneField_ConstrainedValueType` is expected.

Reply:
Thank you for the detailed explanations, it's extremely helpful.